### PR TITLE
feat(taps): Add api costs hook

### PR DESCRIPTION
**Pull request by laurentS**
This PR addresses #348 and adds a simple callback mechanism to the sdk that is:
- totally optional: taps won't see any difference in behaviour (not even logging output) until they implement `calculate_api_request_cost` in their streams.
- flexible: each tap can aggregate cost along any number of dimensions/domains. The only constraint is to provide all dimensions on each request (even if associated cost is 0)
- easy to implement: see https://github.com/MeltanoLabs/tap-github/pull/142
- untested :sweat_smile: I'm not sure what the best way to test it is (the implementation in tap-github kind of does that, but it would be good to cover it in the sdk itself).
- undocumented :see_no_evil: I'll add that once we're all happy with the code

Here is an example of output (at the very end of the tap run) for `tap-github` with the implementation linked above:
```
time=2022-06-08 12:24:42 name=tap-github level=INFO message=Total API costs for stream languages: {'rest': 2, 'graphql': 0, 'search': 0}
time=2022-06-08 12:24:42 name=tap-github level=INFO message=Total API costs for stream repositories: {'rest': 2, 'graphql': 0, 'search': 0}
time=2022-06-08 12:24:42 name=tap-github level=INFO message=Total API costs for stream stargazers: {'rest': 0, 'graphql': 7, 'search': 0}
```

On the downside, the logging above could be improved for better machine handling. `log_api_costs` could be overridden by tap developers for this. And the way I call the logging method at the end feels a bit hacky, but I couldn't think of a better way to do it.

### CODE REVIEW

**Comment by laurentS** [comment]
@aaronsteers @edgarrmondragon it looks like the CI setup is not fully there yet for PRs, the job on python 3.10 (py, external) complains about missing credentials for `tap-google-analytics`, and the `pre-commit` one cannot find poetry. I'm not sure how to fix that last one.

**Comment by edgarrmondragon** [comment]
@laurentS Don't worry about the pre-commit check. The external tests can be retried by an admin (I'll look into making it wait for approval instead of just failing)

File: singer_sdk/streams/core.py
Comment (edgarrmondragon) [inline singer_sdk/streams/core.py:870]:
```suggestion
    def log_sync_costs(self) -> None:
        """Log a summary of Sync costs.

        The costs are calculated via `calculate_sync_cost`.
        This method can be overridden to log results in a custom
        format. It is only called once at the end of the life of
        the stream.
        """
        if len(self._sync_costs) > 0:
            msg = f"Total Sync costs for stream {self.name}: {self._sync_costs}"
            self.logger.info(msg)
```

File: singer_sdk/streams/core.py
Comment (edgarrmondragon) [inline singer_sdk/streams/core.py:80]:
```suggestion
    _sync_costs: Dict[str, int] = {}
```

File: singer_sdk/streams/rest.py
Comment (edgarrmondragon) [inline singer_sdk/streams/rest.py:323]:
```suggestion
            self.update_sync_costs(prepared_request, resp, context)
```

File: singer_sdk/streams/rest.py
Comment (edgarrmondragon) [inline singer_sdk/streams/rest.py:359]:
```suggestion
        call_costs = self.calculate_api_request_cost(request, response, context)
        self._sync_costs = {
            k: self._sync_costs.get(k, 0) + call_costs.get(k, 0)
            for k in call_costs.keys()
        }
        return self._sync_costs
```

File: singer_sdk/tap_base.py
Comment (edgarrmondragon) [inline singer_sdk/tap_base.py:383]:
```suggestion
            stream.log_sync_costs()
```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
@laurentS I've added a few suggestions to rename `api` to `sync` in the base class, since I can imagine a more generic cost calculation for files, databases, etc.

Other points that might be good to start a discussion on:

1. **Which logger should be used**. This is using the tap/stream logger, but I wonder if it makes sense to use a different one so metrics and tap behavior are not mixed in logs (releated to https://github.com/meltano/sdk/issues/383#issuecomment-1154701837)

1. Adding some tests to validate cost calculation in the general case.

**Comment by laurentS** [comment]
> @laurentS I've added a few suggestions to rename `api` to `sync` in the base class, since I can imagine a more generic cost calculation for files, databases, etc.
Ok, makes sense, I've applied your suggestions. Thanks for writing them! I had gone with `api` since it was all network requests, but it's true the requests could be to anything.
 
> Other points that might be good to start a discussion on:
> 
>     1. **Which logger should be used**. This is using the tap/stream logger, but I wonder if it makes sense to use a different one so metrics and tap behavior are not mixed in logs (releated to [#383 (comment)](https://github.com/meltano/sdk/issues/383#issuecomment-1154701837))

From my perspective, this would be much better. Right now, the costs are logged at `info` level, but there is a lot of logging that I have no interest in at that same level. I hesitated to make the costs `warning`s but semantically, that feels wrong. I like your approach of splitting behaviour logging from metrics (though in general I have no use for the general metrics telling me how long each network request took, or the number of records for each of them). I guess we could add something like the `_metric_logging_function` to allow users to set the level they want for various parts of logging (behaviour, metrics, costs?).

>     2. Adding some tests to validate cost calculation in the general case.

I will try to write something for this.

**Comment by laurentS** [comment]
@edgarrmondragon I've updated the code to match your suggestions, and added some testing to validate the summing code.
What do you suggest we do about the logging? If we go with something different as per the [comment](https://github.com/meltano/sdk/issues/383#issuecomment-1154701837) you linked to, should we do this as part of a separate PR?

File: singer_sdk/tap_base.py
Comment (edgarrmondragon) [inline singer_sdk/tap_base.py:384]: Wdyt about moving this to the first loop?

```suggestion
            stream.sync()
            stream.finalize_state_progress_markers()
            stream.log_sync_costs()
```

File: tests/core/test_streams.py
Comment (edgarrmondragon) [inline tests/core/test_streams.py:362]: Thanks for adding this!

**Comment by edgarrmondragon** [comment]
> @edgarrmondragon I've updated the code to match your suggestions, and added some testing to validate the summing code. What do you suggest we do about the logging? 

@laurentS thanks for the updates! I left a single suggestion. Other than that, this looks good to merge 😄.

> If we go with something different as per the [comment](https://github.com/meltano/sdk/issues/383#issuecomment-1154701837) you linked to, should we do this as part of a separate PR?

Let's leave that for another PR, another day 👍

File: singer_sdk/tap_base.py
Comment (laurentS) [reply singer_sdk/tap_base.py:384]: The only reason I did it this way was so all logging would be grouped at the end of the tap run, and not sprinkled around.
Since I mostly grep logs, it doesn't actually matter too much, now that I think of it.
Code would look cleaner without this. Your call :)

Comment (edgarrmondragon) [reply singer_sdk/tap_base.py:384]: @laurentS Yeah, if you don't mind too much and still works for the use case you have in mind, let's keep the single loop 🙂

File: tests/core/test_streams.py
Comment (edgarrmondragon) [inline tests/core/test_streams.py:379]: @laurentS Just to make sure we cover the `log_sync_costs` method:

```suggestion
def test_sync_costs_calculation(tap: SimpleTestTap, caplog):
    """Test sync costs are added up correctly."""
    fake_request = requests.PreparedRequest()
    fake_response = requests.Response()

    stream = RestTestStream(tap)

    def calculate_test_cost(
        request: requests.PreparedRequest,
        response: requests.Response,
        context: Optional[Dict],
    ):
        return {"dim1": 1, "dim2": 2}

    stream.calculate_sync_cost = calculate_test_cost
    stream.update_sync_costs(fake_request, fake_response, None)
    stream.update_sync_costs(fake_request, fake_response, None)
    assert stream._sync_costs == {"dim1": 2, "dim2": 4}

    with caplog.at_level(logging.INFO, logger=tap.name):
        stream.log_sync_costs()

    assert len(caplog.records) == 1

    for record in caplog.records:
        assert record.levelname == "INFO"
        assert f"Total Sync costs for stream {stream.name}" in record.message
```

Comment (edgarrmondragon) [reply tests/core/test_streams.py]: @laurentS file is missing an `import logging` line. That's the last hoop 🙏

Comment (laurentS) [reply tests/core/test_streams.py]: Fixed. Sorry about this. I need to be more careful with the github suggestions system :sweat_smile:

File: singer_sdk/tap_base.py
Comment (laurentS) [reply singer_sdk/tap_base.py]: Well, I messed up here by replying without properly looking at the code. There was another reason that I used 2 loops: the first loop skips all child streams, so the logging only happens with the top level stream (this is what I'm seeing in our logs now).
My bad. I will open a separate PR to revert this change, we do need 2 loops in the end.

Comment (laurentS) [reply singer_sdk/tap_base.py]: The fix is in #752. @edgarrmondragon sorry for messing this up.

**Pull request merged**
